File: search_algorithm/binary_search.cc
#include <iostream>
#include <cmath>
#include <ctime>
#include <cassert>

using namespace std;

template<typename T>

int binary_search(T arr[], int n, T target){

    int l = 0, r = n - 1;  // 在[l...r] 的范围里寻找target

    while (l <= r){    // 当 l == r 时， 区间[l...r] 依然是有效的

        // int mid = (l+r)/2;  // 可能会导致整型溢出
        int mid = l + (r-l)/2;

        if (arr[mid] == target){
            return mid;
        }

        if (arr[mid] < target){
            l = mid + 1;    /* target 在[mid+1 ... r] 中 */
        }
        else{
            r = mid - 1;   /* target 在 [l...mid-1] 中 */
        }
    }
  
    return -1;  
}

File: 查找算法/binary_search.py
# -*- coding:utf-8 -*-

def binary_search(items, item):
    low = 0
    high = len(items) - 1

    while low <= high:

        mid = low + (high-low) // 2
        guess = items[mid]
        if guess == item:
            return mid
        if guess > item:
            high = mid -1
        else:
            low = mid + 1

    return low


def binary_search_high(arr, n, target):
    """在 arr 的 前n个元素内查找target """
    low = 0
    high = n-1

    while low <= high:
        mid = low + (high-low) // 2
        guess = arr[mid]
        if guess == target:
            return mid
        if guess > target:
            high = mid - 1
        else:
            low = mid + 1

if __name__ == "__main__":
    example = [1, 3, 5, 6, 7, 9, 11, 17, 43, 67, 432]

    print(binary_search_high(example,10, 17 ))
